File: src/main/java/no/jstien/bikeapi/utils/ListUtils.java
package no.jstien.bikeapi.utils;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.function.Consumer;

public class ListUtils {
    public static <T> void processBatchwise(List<T> list, int maxBatchSize, Consumer<List<T>> callback) {
        int itemNo = 0;
        int batchItem = 0;
        List<T> buffer = new ArrayList<>();

        while (itemNo < list.size()) {
            buffer.add(list.get(itemNo));

            itemNo++;
            batchItem++;

            if (batchItem >= maxBatchSize) {
                callback.accept(Collections.unmodifiableList(buffer));
                batchItem = 0;
                buffer.clear();
            }
        }

        if (!buffer.isEmpty()) {
            callback.accept(Collections.unmodifiableList(buffer));
            buffer.clear();
        }
    }
}

File: src/main/java/no/jstien/bikeapi/utils/ListUtils.kt
package no.jstien.bikeapi.utils

import java.util.ArrayList
import java.util.Collections
import java.util.function.Consumer

object ListUtils {
    fun <T> processBatchwise(list: List<T>, maxBatchSize: Int, callback: (List<T>) -> Unit) {
        var itemNo = 0
        var batchItem = 0
        val buffer = ArrayList<T>()

        while (itemNo < list.size) {
            buffer.add(list[itemNo])

            itemNo++
            batchItem++

            if (batchItem >= maxBatchSize) {
                callback(Collections.unmodifiableList(buffer))
                batchItem = 0
                buffer.clear()
            }
        }

        if (!buffer.isEmpty()) {
            callback(Collections.unmodifiableList(buffer))
            buffer.clear()
        }
    }
}
